finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+    cooldown: time::Duration,
+    last_broadcast: HashMap<NodeIndex, time::Instant>,
+    last_response: HashMap<NodeIndex, time::Instant>,
+    requested_authentications: HashMap<NodeIndex, HashSet<NodeIndex>>,
+    next_query: NodeIndex,
+}
+
+fn authentication_broadcast(authentication: Authentication) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::AuthenticationBroadcast(authentication),
+        DataCommand::Broadcast,
+    )
+}
+
+fn request(
+    missing_authorities: Vec<NodeIndex>,
+    authentication: Authentication,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Request(missing_authorities, authentication),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+fn response(
+    authentications: Vec<Authentication>,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: time::Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(
+        &mut self,
+        handler: &SessionHandler,
+    ) -> Vec<(DiscoveryMessage, DataCommand)> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if missing_authorities.len() * 3 > 2 * node_count.0 {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < 2 {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
```

Comment:
Ouch, I've been thinking a lot about this. If we want to do this this way it would probably have to be behind an `Arc<Mutex<..>>` or expose a separate interface for accessing the internal handler. Both are kinda ugly, but so is the current solution... I ended up choosing this one, and assuming I would keep pairs `(Discovery, SessionHandler)` in the object that will end up using this, but it should be easy to convince me that this is not the correct choice.

Reply:
I would go probably for the `Arc<Mutex<..>>` solution (btw, maybe it deserves a shortcut type), since I really don't like a bunch of methods with the same argument passed all around. However, unwrapping it everywhere would be awful and maybe some setter would be necessary (for changing `SessionHandler`?). 
I would leave the decision to the second reviewer, @Yithis :stuck_out_tongue_closed_eyes:

Reply:
So sorry @pmikolajczyk41, but I'm with @timorl here :sweat_smile:

Reply:
sure, thus resolving the conversation